polygeist/lib/Dialect/Polygeist/Transforms/DetectReduction.cpp

```diff
@@ -73,35 +73,174 @@ class ReductionOp {
+public:
+  LoopReductionIter(const LoopReductionIter &) = delete;
+  LoopReductionIter(LoopReductionIter &&) = delete;
+  LoopReductionIter &operator=(const LoopReductionIter &) = delete;
+  LoopReductionIter &operator=(LoopReductionIter &&) = delete;
+
+protected:
   Pass::Statistic &numReductionsDetected;
 
-public:
-  AffineForReductionIter(MLIRContext *context,
-                         Pass::Statistic &numReductionsDetected)
-      : OpRewritePattern<AffineForOp>(context),
-        numReductionsDetected(numReductionsDetected) {}
+  LoopReductionIter(Pass::Statistic &numReductionsDetected)
+      : numReductionsDetected(numReductionsDetected) {}
+  virtual ~LoopReductionIter() = default;
+
+  virtual Block::BlockArgListType
+  getRegionIterArgs(LoopLikeOpInterface Loop) const = 0;
 
+  virtual unsigned getNumRegionIterArgs(LoopLikeOpInterface Loop) const = 0;
+
+  virtual void cloneFilteredTerminator(
+      Operation *Terminator,
+      const SmallVectorImpl<ReductionOp> &ReductionOps) const = 0;
+
+  virtual LoopLikeOpInterface
+  createNewLoop(LoopLikeOpInterface Loop, SmallVectorImpl<Value> &NewIterArgs,
+                PatternRewriter &Rewriter) const = 0;
+
+  LogicalResult matchAndRewrite(LoopLikeOpInterface Loop,
+                                PatternRewriter &Rewriter) const {
+    SmallVector<ReductionOp> ReductionOps;
+    WalkResult Result = collectCandidates(Loop, ReductionOps);
+    if (Result.wasInterrupted() || ReductionOps.empty()) {
+      LLVM_DEBUG({
+        llvm::dbgs() << "No reduction found\n";
+        llvm::dbgs() << "------------------------------------------------\n";
+      });
+      return failure();
+    }
+
+    LLVM_DEBUG({
+      llvm::dbgs() << "\nFound " << ReductionOps.size() << " reduction(s):\n";
+      for (ReductionOp &Op : ReductionOps)
+        llvm::dbgs() << Op << "\n";
+      llvm::dbgs() << "\n";
+    });
+
+    // Move the load outside the loop (recall that the load is loop invariant).
+    // The load result is passed to the new loop as an iter argument.
+    SmallVector<Value, 4> NewIterArgs(getRegionIterArgs(Loop));
+    Rewriter.setInsertionPoint(Loop);
+    LLVM_DEBUG(llvm::dbgs() << "New Loop:\n");
+    for (const ReductionOp &Op : ReductionOps) {
+      Operation *MovedLoad = Rewriter.clone(*Op.getLoad());
+      LLVM_DEBUG(llvm::dbgs() << *MovedLoad << "\n");
+      NewIterArgs.push_back(MovedLoad->getResult(0));
+    }
+
+    // Create the new loop.
+    LoopLikeOpInterface NewLoop = createNewLoop(Loop, NewIterArgs, Rewriter);
+
+    // Remove the load operation inside the new loop.
+    size_t ArgNo = 0, OrigNumRegionArgs = getNumRegionIterArgs(Loop);
+    for (const ReductionOp &Op : ReductionOps) {
+      Op.getLoad()->getResult(0).replaceAllUsesWith(
+          NewLoop.getLoopBody().getArguments()[ArgNo + OrigNumRegionArgs + 1]);
+      Rewriter.eraseOp(Op.getLoad());
+      ++ArgNo;
+    }
+
+    Region &NewBlock = NewLoop.getLoopBody(), &OldBlock = Loop.getLoopBody();
```

Comment:
```suggestion
    Block *NewBlock = &NewLoop.getLoopBody().front(), *OldBlock = &Loop.getLoopBody().front();
```
or name them `NewRegion` and `OldRegion`.